qi-doc/scribblings/field-guide.scrbl

```diff
@@ -31,7 +31,37 @@ A journeyman of one's craft -- a woodworker, electrician, or a plumber, say -- a
 
 @subsection{Be Intentional About Effects}
 
-Qi encourages a style that avoids "accidental" effects. A flow should either be pure (that is, it should be free of "side effects" such as printing to the screen or writing to a file), or its entire purpose should be to fulfill a side effect. It is considered inadvisable to have a function with sane inputs and outputs (resembling a pure function) that also performs a side effect. It would be better to decouple the effect from the rest of your function (@seclink["Use_Small_Building_Blocks"]{splitting it into smaller functions}, as necessary) and perform the effect explicitly via the @racket[effect] form, or otherwise escape from Qi using something like @racket[esc] (note that @seclink["Identifiers"]{function identifiers} used in a flow context are implicitly @racket[esc]aped) in order to perform the effect. This will ensure that there are no surprises with regard to @seclink["Order_of_Effects"]{order of effects}.
+Qi encourages a style that avoids "accidental" effects.
+
+In functional programming, "effects" refer to anything that the function does that is not captured in its inputs and outputs. This could include things like printing to the screen, writing to a file, or mutating a global variable.
+
+A @tech{flow} should either be pure (that is, free of such side effects), or its entire purpose should be to fulfill a side effect. It is considered inadvisable to have a function with sane inputs and outputs (resembling a pure function) that also performs a side effect. It would be better to decouple the effect from the rest of your function (@seclink["Use_Small_Building_Blocks"]{splitting it into smaller functions}, as necessary) and perform the effect explicitly via the @racket[effect] form, or otherwise escape from Qi using something like @racket[esc] (note that @seclink["Identifiers"]{function identifiers} used in a flow context are implicitly @racket[esc]aped) in order to perform the effect. This will ensure that there are no surprises with regard to @seclink["Order_of_Effects"]{order of effects}.
```

Comment:
I don't like "considered inadvisable" here. I think it would be better to make the reasons explicit up front. Something like, "Qi may reorder operations not marked as effects, so it's better to separate effects from other computations to ensure they work as expected".

I don't think I understand how `esc` and `effect` relate to suppressing effect re-ordering. Are you promising that neither will ever be reordered, or are you promising that `effect` will never be reordered but also suggesting that pairs of effects inside the same `esc` in Racket code will of course not be re-ordered relative to each other, but not making any promises re: pairs of effects in separate `esc` forms?

Reply:
Yes, something like the latter. I think we are promising:

- If you use `esc`, anything inside that will be untouched (but it may still be reordered wholesale at a higher level)
- If you use `effect`, the effect will happen along with the annotated flow, whenever that happens (but the whole side-effecting flow `(effect f g)` may itself be moved around)

To take an example, with this normalization rule:

```
[(thread _0 ... (pass f) (amp g) _1 ...)
 #'(thread _0 ... (amp (if f g ground)) _1 ...)]
```

It would rewrite:

```
(~> (pass (ε displayln positive?)) (amp sqr)) --> (amp (if (ε displayln positive?) sqr ⏚))
```

Here, the effect stays with the flow it annotated, but it happens at a different time.

So what we are promising may be some kind of "effect locality" rather than any particular order of effects. Tbh I'm not sure if this is a good way to think about it and what exactly our guarantees imply. I will think about it some more.

On a side note, the former of these bullets (re: `esc`) makes me feel that matching literal uses of `racket/list`'s `map`, `filter`, etc. is probably not the right long term thing, and that we'd probably want a `qi/list` language that actually includes `map` and `filter` forms. This would ensure that we do not cross the `esc` boundary to do optimizations.

Reply:
So the only way to really guarantee order of effects is to wrap each flow in `effect`? Maybe? [I don't think this will matter most of the time to most people, and it suggests that highly effectful programs benefit from a more "direct" imperative style.]

Reply:
I'm not sure if wrapping every flow with `effect` would guarantee an order, except that doing so could mean that no compiler pattern would match (but in this case, if there are just a lot of empty `effect` forms, we would likely normalize that away in any case, like `(effect ground flo)` should, I think, be rewritten to `flo`).

Here's a first attempt at formalizing our guarantees about effects:

For two flows `f` and `g`, we could define a relation "g is downstream of f" as the outputs of `f` are used, either directly or transitively, as inputs to `g`.

Then, our guarantees about effects are:

1. Any `effect`s on `f` will occur before any effects on `g`.
2. `effect`s on any flow `φ` will happen when `φ` is called.

That is, we guarantee that an `effect` will never be separated from the flow it annotates (what we could call "locality"), though _when_ this is invoked is not guaranteed, aside from point (1) above.

Do we feel this is a useful characterization?

Reply:
This might seem silly, but do we optimize either of the flows within `(effect f g)`?

Reply:
Yes, we would optimize both.